src/data/copying_prob_src.rs

```diff
@@ -0,0 +1,154 @@
+extern crate rand;
+use rand::distributions::{Range, IndependentSample};
+use af;
+use af::{Array, Dim4, DType};
+use std::cell::{RefCell, Cell};
+
+use initializations;
+use utils;
+
+use data::{Data, DataSource, DataParams, Normalize, Shuffle};
+
+pub struct CopyingProblemSource {
+    pub params: DataParams,
+    pub iter: Cell<u64>,
+    pub offset: Cell<f32>,
+    pub bptt_unroll : u64, 
+    pub seq_size : u64,
+}
+
+impl CopyingProblemSource {
+    pub fn new(input_size: u64, batch_size: u64, seq_size: u64
+               , bptt_unroll: u64, dtype: DType, max_samples: u64) -> CopyingProblemSource
+    {
+        let input_dims = Dim4::new(&[batch_size, input_size, bptt_unroll, 1]);
+        //let output_dims = Dim4::new(&[batch_size, output_size, bptt_unroll, 1]);
+        let train_samples = 0.7 * max_samples as f32;
+        let test_samples = 0.2 * max_samples as f32;
+        let validation_samples = 0.1 * max_samples as f32;
+        CopyingProblemSource {
+            params : DataParams {
+                input_dims: input_dims,
+                target_dims: input_dims,
+                dtype: dtype,
+                normalize: false,
+                shuffle: false,
+                current_epoch: Cell::new(0),
+                num_samples: max_samples,
+                num_train: train_samples as u64,
+                num_test: test_samples as u64,
+                num_validation: Some(validation_samples as u64),
+            },
+            iter: Cell::new(0),
+            offset: Cell::new(0.0f32),
+            bptt_unroll: bptt_unroll,
+            seq_size: seq_size,
+        }
+    }
+
+    fn generate_input(&self, batch_size: u64, input_size: u64, bptt_unroll: u64, seq_size: u64) -> Array {
+
+        let between = Range::new(0,input_size-1);
+        let mut rng = rand::thread_rng();
+        
+        let mut vec_total = Vec::new();
+        let vec = vec!(0f32; input_size as usize);
```

Comment:
Please do not call a vector vec. This is confusing with rust's notation of the macro `vec!`. Descriptive names are useful in debugging later.